docs/src/design/expressions.md

```diff
@@ -0,0 +1,130 @@
+# Expressions
+
+Expressions are *domain specific language (DSL)* built by Mapbox for
+vector styles. Mapbox Vector Style is used in Mapbox Vector Tiles.
+Mapbox Vector Tiles is a vector tile specification initiated by Mapbox
+which was later widely adopted by the geospatial community.
+
+To recap, Mapbox Vector Styles have 3 components:
+
+1.  Sources
+
+2.  Layers
+
+3.  Expressions
+
+Sources contain geospatial features, in GeoJSON format. It's the data to
+be visualized on the map. Layers contains data from a source. This is
+where we expressions kick in. Expressions define how the data from a
+source will be painted in a layer following a style. For example, a
+heatmap, require the ability to paint features in different zoom levels
+with different colors. Expressions facilitate that. The rendering
+depends on the style of the layer along with pitch, bearing, and zoom of
+the map. This was called *Data Driven Styling (DDS)*. Another option
+that was used was to completely change the style in run time to achieve
+the same outcome[^17].
+
+The desire of being able to render a layer in different zoom levels
+differently based on data drove the birth of *expressions*. To
+summarize, expressions do the following:
```

Comment:
I would phrase this more as "common uses of expressions are:" because, any property in the mapbox style that takes an expression, you can do something custom/interesting there, beyond the specific use cases mentioned here.